File: Jongryul/09/14002.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";

const input = fs.readFileSync(filePath).toString().trim().split("\n");
const n = Number(input[0]);
const arr = input[1].split(" ").map(Number);
const dp = Array(n).fill(1);
const parent = Array.from({ length: n }, (_, i) => i);

for (let i = 0; i < n; i++) {
  for (let j = i + 1; j < n; j++) {
    if (arr[i] < arr[j] && dp[i] + 1 > dp[j]) {
      dp[j] = dp[i] + 1;
      parent[j] = i;
    }
  }
}

let mx = dp[0];
let max_idx = 0;

for (let i = 1; i < n; i++) {
  if (dp[i] > mx) {
    mx = dp[i];
    max_idx = i;
  }
}

console.log(mx);

const rlt = [];
let p = max_idx;
while (true) {
  rlt.unshift(arr[p]);
  if (arr[p] === arr[parent[p]]) {
    break;
  }
  p = parent[p];
}
console.log(rlt.join(" "));

File: Season 1/07th_Competition/Jongryul/09/14002.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
dp = [1] * n
parent = [i for i in range(n)]

for i in range(n):
    for j in range(i+1, n):
        if arr[i] < arr[j]:
            if dp[i] + 1 > dp[j]:
                dp[j] = dp[i]+1
                parent[j] = i

mx = dp[0]
max_idx = 0
for i in range(n):
    if dp[i] > mx:
        mx = dp[i]
        max_idx = i

rlt = []
p = max_idx
while True:
    rlt.append(arr[p])
    if arr[parent[p]] == arr[p]:
        break
    p = parent[p]

print(mx)
print(*reversed(rlt))